packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3984,14 +3994,24 @@ class InheritedElement extends ProxyElement {
   final Set<Element> _dependents = new HashSet<Element>();
 
   @override
-  void _updateInheritance() {
+  void update(covariant InheritedWidget newWidget) {
+    if (widget.inheritedType != newWidget.inheritedType) {
+      _child = updateChild(_child, null, slot);
```

Comment:
why null?

Reply:
This line is the sadness in this PR.  Previously, if you changed the type of inherited widget, then we'd blow away the subtree because the runtimeType of the widget itself changed.  This line does the same thing when you get a new widget with a different inheritedType.

The trouble is that the subtree below this point has inherited widget maps that refer to the old type and need to be updated.  There isn't really a good way to update them without just recreating the subtree.

Reply:
oh, i see, you're forcing the widget to go through the deactivate/activate cycle so that its _dependencies are cleared?

Reply:
maybe mention in the docs somewhere that if the type changes from instance to instance, it's going to be relatively expensive (rebuild the entire subtree), as if the widget had changed identity?